### PR TITLE
feat(common): initial use of lerna (in-repo package links only)

### PR DESCRIPTION
**Pull request by jahorton**
This PR marks the initial use of [`lerna`](https://github.com/lerna/lerna) within our repository.  At time, we'll take things nice and slow, relying only on the `lerna bootstrap` command, as that's the part we actually need right now.  (We can always experiment with more at a later point.)

`lerna bootstrap` analyzes the packages that _we set_ within the base-level `lerna.json` file.  If any of those packages are dependencies of one of the others, `lerna bootstrap` will set up the `npm link`s for us... and afterwards it will automatically `npm install` everything else.

Should there be an error trying to install a package, say, if it were missing:

missing package trial run

All in all, it seems pretty simple to use, at least at this minimal level.  There are other features we may wish to consider eventually - the first being [dev-dependency pooling](https://github.com/lerna/lerna#common-devdependencies) - but we can cross those bridges later.

So... now welcoming our [lerna](https://lerna.js.org/) overlords.  No need to worry about them auto-publishing or anything; control is firmly in our hands.

Please note:  most of the 'big' changes that appear to be present are due to package-lock.json (automated) entries.

### CODE REVIEW

File: common/predictive-text/package.json [hunk 1]
```diff
@@ -1,14 +1,13 @@
 {
-  "name": "keyman-lmlayer-prototype",
-  "version": "0.0.0",
+  "name": "keyman-lmlayer",
+  "version": "14.0.0",
```
Comment (eddieantonio) [inline common/predictive-text/package.json:3]: 👍

File: common/predictive-text/package.json [hunk 2]
```diff
@@ -1,14 +1,13 @@
 {
-  "name": "keyman-lmlayer-prototype",
-  "version": "0.0.0",
+  "name": "keyman-lmlayer",
```
Comment (eddieantonio) [inline common/predictive-text/package.json:2]: Haha, it's been a prototype all this time, and it's been released in two major versions! 😂

File: common/predictive-text/package.json [hunk 3]
```diff
@@ -1,14 +1,13 @@
...
+  "version": "14.0.0",
   "description": "Keyman/Predictive Text integration layer",
   "main": "build/index.js",
   "scripts": {
     "karma": "karma",
     "mocha": "mocha",
     "test": "bash ./unit_tests/test.sh",
-    "tsc": "tsc",
-    "install": "npm link @keymanapp/lexical-model-types"
```
Comment (eddieantonio) [inline common/predictive-text/package.json:11]: This actually fixes an issue in modern versions of Node/npm, so I welcome this addition! 👍

File: web/package.json [hunk 4]
```diff
@@ -49,6 +49,7 @@
     "@types/node": "^11.9.4",
     "es6-shim": "^0.35.5",
     "eventemitter3": "^4.0.0",
+    "keyman-lmlayer": "^14.0.0",
```
Comment (eddieantonio) [inline web/package.json:52]: 👍

**Comment by eddieantonio** [comment]
> I've checked this out on my work machine, but I was unable to get it building. It seems that the `lexical-model-types` package is not linked in `common/predictive-text`:
> 
> ```
> ~/Work/keyman/common/predictive-text (feat/common/init-lerna) $ ./build.sh
> lerna notice cli v3.20.2
> lerna info Bootstrapping 3 packages
> lerna info Installing external dependencies
> lerna info Symlinking packages and binaries
> lerna success Bootstrapped 3 packages
> 
> > keyman-lmlayer@14.0.0 tsc /Users/santoseadmin/Work/keyman/common/predictive-text
> > tsc "-p" "./worker/tsconfig.json"
> 
> message.d.ts:23:22 - error TS6053: File '/Users/santoseadmin/Work/keyman/common/predictive-text/node_modules/@keymanapp/lexical-model-types/index.d.ts' not found.
> 
> 23 /// <reference path="./node_modules/@keymanapp/lexical-model-types/index.d.ts" />
> ```
> 
> I haven't investigated this further, but just FYI.

Update! I fixed this by running:

```sh
lerna add --dev @keymanapp/lexical-model-types common/predictive-text
```

...in the repository root. This changed the dependencies in `common/predictive-text/package.json` accordingly.

**Comment by jahorton** [comment]
Ah, I probably forgot to actually add it as a dependency to the package.  `lerna` reads those for its bootstrapping.  It doesn't remove previously set links unless they conflict with a link it actively wants to setup.  I'll make sure that gets put in.

File: common/predictive-text/package.json (hunk 1 above)
Comment (mcdurdin) [reply common/predictive-text/package.json:3]: Is this version number static? Or is it kept in sync with VERSION.md?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "root",
```
Comment (mcdurdin) [inline package.json:2]: Is `"root"` the most appropriate name? Or should it be `"keyman"`? I am unsure!

File: common/predictive-text/package.json (hunk 2 above)
Comment (mcdurdin) [reply common/predictive-text/package.json:2]: It occurs to me that we should probably be going with `@keymanapp/lexical-model-layer`. to match `@keymanapp/lexical-model-types` and `@keymanapp/lexical-model-compiler`. There are some additional namespacing concerns, that arise as we use npm more, such as ... perhaps we should be `@keymanapp/developer-lexical-model-compiler` instead of `@keymanapp/lexical-model-compiler` and `@keymanapp/common-lexical-model-types` etc so that the dependency map doesn't become too tangled in our own minds!

Comment (jahorton) [reply common/predictive-text/package.json:2]: I think I'll hold off on those latter points for now.

At what point do we not namespace things?  I just checked [Sentry's setup](https://github.com/getsentry/sentry-javascript#supported-platforms) as a reference, and it seems like all of their 'current' JS repos are all 'namespaced' like you suggest.

So, should I do this with `@keymanapp/keyman`, `@keymanapp/keyboard-processor`, etc?

File: lerna.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "packages": [
+    "common/predictive-text",
+    "common/lexical-model-types",
+    "web"
+  ],
+  "version": "14.0.0"
```
[the remark this answers is not in the document]
Comment (jahorton) [reply lerna.json:7]: Manually.  We're not publishing, so there's no auto-update of version numbers.

That said, the version number actually does matter a bit; if Lerna can't find a matching version for a package within the repo, it won't set up the internal links.  So, I boosted the packages it'll manage to 14.0.0 so that there's at least a simple baseline for this.

For the future:  `lerna publish`, should we ever start to use that to formally publish our packages, apparently can auto-submit PRs to increment version numbers.

File: package.json (hunk 5 above)
Comment (jahorton) [reply package.json:2]: That's `lerna`'s package.json, basically a 'pseudo-package'. It's used to facilitate its NPM operations.

I could experiment and see if it'll function correctly should I alter it.  I figure `lerna` operations may search the file-tree to find the 'root' package, so I didn't want to take a chance with that yet.

File: common/predictive-text/package.json (hunk 1 above)
Comment (jahorton) [reply common/predictive-text/package.json:3]: Static, for now.

**Comment by jahorton** [comment]
> It's pretty difficult to wrap my mind around the changes across two separate PRs for `lerna`. I think we are all agreed that the non-global version of `lerna` is better.
>
> tl;dr: can you just merge that branch into this one and I can re-review on the final intended PR?

Now that it's explicit, sure.

**Comment by jahorton** [comment]
> FWIW, I am coming to find this to be an issue for reviewing chained Pull Requests in general. The mental load to review is much higher because the reviewer has to figure out the union of changes in order to understand the final impact of the PR. The reviewer can choose to review on a diff of the final product -- but then have to transfer their comments from that view to the PR view, or they can flip back and forth between the base PR and the child. Or they can ignore the base PR and review the child PR completely independently. All of these approaches have significant drawbacks of course and generally lead to lower quality reviews 😢.

For this pair, I can definitely understand; there's a fair bit of re-altered code in the other one.  That 'other one' was really meant more as a 'patch' of sorts, since I didn't have an explicit "go-ahead" for its approach and a different suggestion had been blocked.  Sorry I forgot to mark it with a "go ahead and merge it if you'd prefer."

-----

In general though... isn't it kind of a "pick your poison" thing?  Either tons of changes get done all at once (like #496, with +3924, -3201 or like #1519, with +5,722 −5,208) or we break things down into smaller chunks like most of this web-core refactoring has been.

A very common complaint of ours with the larger ones, even when more in the (+500, -500)-ish range, was that the sheer scale of those PRs made it impossible to thoroughly vet them, or caused people to skim and not pay attention.  Heck, just the other day on #2986, we had complaints when a PR was in the +250, -150 range.  (Granted, there _was_ a better way of doing it, as the PR review proved, and the review did come out quite high-quality.)  You even [gave a +1](https://github.com/keymanapp/keyman/pull/2986#issuecomment-614741632) to one of the comments to that effect, probably regarding the following line:

> Ideally, a refactor PR should be just that: one refactor. Extract superclass is a big move, and ideally, this should be its own PR.

Given such concerns, I've been making a lot of these smaller ones, generally trying to approach that "one refactor" idea but taking it just a bit loosely.  I've been aiming for the scale of "one _conceptual_ refactor" or "one design change" - there may be more than just one set of changes per PR, but I aim to keep them all _highly_ related and focused.  Generally, the 'other changes' are necessary to facilitate the PR's true focus.  Going any finer grained than that would result in a near-explosive PR count... and a bit of a local-branch management nightmare.  (We're already approaching my branch management limits, as evidenced by one recent merge-accident of mine.)

Are you saying that _most_ of the recent web-core PRs have been broken down too far, or just that a few (like this one) were?

> This is particularly the case for big refactor PRs. Those are hard enough to review well without the added complication of multiple dependent PRs! I appreciate that changing the way we do this may slow you down so I'd like to consider alternatives. Something to think about.

I feel like the usual issue with "big refactor PRs" is that there are so many conceptual changes that it's hard to track everything.  That's why I've aimed for the "one design change" idea mentioned above; I've been designing these PRs to automatically highlight the most important changes at a conceptual level so that the overall refactor, from beginning to end, has an easily-followable approach.

Will some small specifics be missed?  Almost certainly... but they'd _absolutely_ be missed in a giant PR.  (I've already caught a few that were missed, fixing them as part of #2994.)  Honestly, though... breaking things down to this level has made it easier to find the bugs when they appear.  The explicitness involved in this process has made it pretty easy to reference which set of changes overlooked something when I've needed to fix something.

With the larger PRs, we definitely run a risk of _big_ elements being missed entirely.  A great example can be found at #2976 in [this comment](https://github.com/keymanapp/keyman/pull/2976#pullrequestreview-394103775) and from its ensuing conversation at the time.  We technically added the `com` global (due to how TS does namespacing) in 10.0, it's been utilized in unit tests multiple times since then, and concerns regarding that were only _just_ brought up during 14.0 alpha!

> I appreciate that changing the way we do this may slow you down so I'd like to consider alternatives. Something to think about.

Aside from the alternative of "just don't refactor or change anything", I'm at a bit of a loss here.  Got any examples of well-done PR sequences (or PRs within sequences) versus ones that were worse, like this one and #3012?  That might help me identify _something_ to consider.

File: lerna.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "packages": [
+    "common/predictive-text",
+    "common/lexical-model-types",
+    "web"
+  ],
```
Comment (jahorton) [inline lerna.json:6]: Note:  this array is managed _**manually**_.

`lerna` will not automatically add packages to this list.  Anything _**we**_ add here, `lerna` will manage, but it will leave other packages alone.

Comment (darcywong00) [reply lerna.json:6]: Would `web/README.md` be the place to document how we manually manage packages with lerna?
Along with updating new build options.

For the 🚌 factor

Comment (jahorton) [reply lerna.json:6]: I'll add something to the documentation for `verify_npm_setup` in its source script for this; that'd be a far better place.

File: common/predictive-text/package.json (hunk 2 above)
Comment (jahorton) [reply common/predictive-text/package.json:2]: I'll go with `@keymanapp/language-modeling-layer` - there's the chance we might eventually incorporate other language-modeling aspects.  That, and I think "language modeling" was the origin of the 'lm' abbreviation anyway.

Comment (mcdurdin) [reply common/predictive-text/package.json]: > At what point do we not namespace things?

Uh ... as I said in the next line: we have `lexical-model-compiler` not `developer-lexical-model-compiler`. Given the lexical model compiler is a part of Developer, it makes sense to namespace it accordingly. But no rush on that.

> I'll go with `@keymanapp/language-modeling-layer`

Why not `lexical-model-layer` so it matches `lexical-model-compiler`? What you've proposed just adds more terms and confusion and breaks the mental link between the components. We should be naming according to current use (esp. since you noted 'chance', 'might', 'eventually' which sounds pretty doubtful!)

**Comment by mcdurdin** [comment]
> Aside from the alternative of "just don't refactor or change anything", I'm at a bit of a loss here. Got any examples of well-done PR sequences (or PRs within sequences) versus ones that were worse, like this one and #3012? That might help me identify _something_ to consider.

tl;dr: we may need to consider a slower pace of refactoring in order to maintain quality/stability.

I think the core issue here is that you have a mental map of where you are going with a refactor that the reviewers just don't have -- and will never have to the same level of detail that you have. So whether we go with a single huge refactor PR, or a bunch of smaller ones, there's still going to be that impedance. The document you wrote up on the changes helps -- but the reviewers are still going to have to load this into their memory before reviewing so that's a challenge.

On top of that, Javascript (especially on web platform) is so very fragile due to lack of adequate encapsulation and a very broad scope of interactions. It requires a rigorous approach from the start to ensure that published APIs don't shift over time. That of course is something we have never really had with KeymanWeb. We've seen multiple times that things have broken when we were not fully aware of how KeymanWeb was being used in all contexts.

Lastly, the codebase for KeymanWeb in particular has definitely suffered from multiple people working on it in isolation over time together with rapidly shifting web standards/fads. The state of the art when KeymanWeb was first built was jQuery, and even that was still very new then.

So there are many reasons why refactoring KMW is always going to be challenging.

I don't necessarily have a good answer here. While refactoring to improve code quality and maintenance is a good thing to do, it does introduce a mountain of risk which I wish we could avoid. Insufficient code review contributes to that.

The ideal refactor PR moves a single function from one place to another, or renames a single variable. The change moves from a 'known good' state to another 'good' state. The consequences of those changes are much easier to reason about and test against.

None of this is intended to be a criticism of you. This is about trying to figure out how we can improve the quality of our code review process and work effectively as a massively resource-constrained team. Some of my Keyman Desktop PRs suffer from similar issues, and I have the advantage of having a full history of the project and don't need to worry quite as much about 3rd party dependencies on Keyman Desktop.

I think we're going to have to run on this tightrope for a while longer -- any choice we make has downsides and the choice we've come to here is probably leaning towards accepting some instability in order to move forward rather than backward.

Keep on keeping on ;-)

File: common/predictive-text/package.json (hunk 1 above)
Comment (jahorton) [reply common/predictive-text/package.json:3]: I just had a major breakthrough on this.

`lerna version --amend <new-version-here> --yes --force-publish`

- [`--amend`](https://github.com/lerna/lerna/tree/master/commands/version#--amend): make changes locally, do not commit.  (Docs even state this is intended for CI)
- [`--yes`](https://github.com/lerna/lerna/tree/master/commands/version#--yes): do not prompt, assume we'll say 'yes' for all prompts.
    - so, this can be safely omitted in non-CI environments; useful for testing/learning the command
    - I'm... fairly certain the example is a copy from the `lerna publish --yes` documentation.
- [`--force-publish`](https://github.com/lerna/lerna/tree/master/commands/version#--force-publish): do not check `git diff` - assume every package changed.  Doesn't actually publish.
    - Not sure exactly why it's named this; there's no matching option on `lerna publish`.
    - In isolation, `lerna version` likes to automatically `git push` the changes to the repository.  So, "publish" here would be "to Github", not "to `npm`."

lerna versioning

Note:
- `Skipping git push`
- `Skipping releases`

----

So, once `lerna` hits `master`, it may be worth adding this call to our daily version-update Action.  We can finally have auto-updated version info in our repo's package.json files!  (Well, the ones managed by `lerna`, anyway.)

Comment (mcdurdin) [reply common/predictive-text/package.json:3]: Good find!

But oh my gosh I really dislike the opinionated my-way-or-the-highway defaults of NPM ... having to specify --no-git-tag-version to disable default functionality is just SO BACKWARDS!

Comment (jahorton) [reply common/predictive-text/package.json:3]: Yeah, I feel you.  _This_ one's on `lerna`; it tries to do everything out-of-the-box for version management.

I've now added an appropriate script definition to the base package:  I've converted the above into a simple `npm run version 14.xx.xx` call at the repo's base.

**Comment by jahorton** [comment]
And one minor retool later, I figured out how to install `lerna` just the once, at the repo's base.  Our build.sh scripts will simply call it from the base package, but I also figured out how to, uh, trick `npm run` into letting us call `lerna`, with appropriate (as in, kinda excessive) escapes from each project in case we want that.  I've added some documentation on that to shellHelperFunctions.sh and have setup the KMW and lm-layer package.json files appropriately.

(Extra benefit:  only one set of serious package-lock.json changes!)

This helps prevent a lot of multiple-install noise, as `lerna` has numerous dependencies we don't otherwise use.  Technically, we should also be able to move a fair few packages higher up in the hierarchy, but I think experimenting with that is best left to a separate PR.

File: .gitignore [hunk 8]
```diff
@@ -258,3 +258,5 @@ resources/environment.sh
 
 # Copy of the master VERSION.md used for Linux packaging
 /common/core/desktop/VERSION.md
+
+**/node_modules/**
```
Comment (mcdurdin) [inline .gitignore:262]:
```suggestion
**/node_modules/
```

File: web/bulk_rendering/build.sh [hunk 9]
```diff
@@ -1,7 +1,14 @@
 #! /bin/bash
 # 
 # Compile the KeymanWeb bulk-renderer module for use with developing/running engine tests.
-#
+
+## START STANDARD BUILD SCRIPT INCLUDE
+# adjust relative paths as necessary
+THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
+KEYMAN_ROOT="$(dirname "$THIS_SCRIPT")/../.."
```
Comment (mcdurdin) [inline web/bulk_rendering/build.sh:8]: I'd prefer `KEYMAN_ROOT` to only be set by `build-utils.sh` -- even though this looks slightly less clean for the build-utils.sh call: this just helps to validate that this variable is 'owned' by `build-utils.sh`. After calling `build-utils.sh`, the subsequent lines should be able to call it without issue. Technically, we should have the `shellHelperFunctions.sh` include outside the `STANDARD BUILD SCRIPT INCLUDE` section... but that I'm less fussy about.

File: common/predictive-text/package.json [hunk 10]
```diff
@@ -1,25 +1,25 @@
 {
-  "name": "keyman-lmlayer-prototype",
-  "version": "0.0.0",
+  "name": "@keymanapp/language-modeling-layer",
```
Comment (mcdurdin) [inline common/predictive-text/package.json:2]: Still hoping for `@keymanapp/lexical-model-layer` in order to match with the types and compiler names etc.

File: web/source/build.sh [hunk 11]
```diff
@@ -6,7 +6,9 @@
 ## START STANDARD BUILD SCRIPT INCLUDE
 # adjust relative paths as necessary
 THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
-. "$(dirname "$THIS_SCRIPT")/../../resources/build/build-utils.sh"
+KEYMAN_ROOT="$(dirname "$THIS_SCRIPT")/../.."
```
Comment (mcdurdin) [inline web/source/build.sh:9]: Ditto on `KEYMAN_ROOT` variable from earlier comment.

File: web/source/build_dev_resources.sh [hunk 12]
```diff
@@ -3,7 +3,14 @@
 # Compiles development-related KeymanWeb resources for use with developing/running tests.
 #   - the Recorder module (for engine tests)
 #   - the DOM module (for touch-alias and element-interface tests)
-#
+
+## START STANDARD BUILD SCRIPT INCLUDE
+# adjust relative paths as necessary
+THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
+KEYMAN_ROOT="$(dirname "$THIS_SCRIPT")/../.."
```
Comment (mcdurdin) [inline web/source/build_dev_resources.sh:10]: Ditto on `KEYMAN_ROOT` variable.

**Comment by jahorton** [comment]
That last tidbit - using the single install of `lerna` at the repo's base - should temporarily resolve the issues; the specific pathname-length issue we ran into arose due to installing `lerna` as a dev-dependency of each project's package.  Neither the `lerna` package nor its unique dependencies will show up at that depth any longer.

That said, we should still be wise when planning our module structure - making too deep a local dependency tree would likely cause the issue to arise again.  We've talked about possibly making micro-packages/modules... I think that as long as they remain "leaves" on the dependency tree, rather than internal nodes, we should be fine.  But, if we manage that poorly, the pathname-length issue may arise again without further changes.

File: common/predictive-text/package.json (hunk 10 above)
Comment (jahorton) [reply common/predictive-text/package.json:2]: I'm going to need some convincing on that.  Thoughts, @eddieantonio?

My perspective was that lexical models are a _part of_ the lm-layer, not the entirety.  During our original design process, we noted the potential for other types of language-models to be involved in the process.  "lexical models" are a strict subset of "language models".  If we start involving any sort of contextual model, say... bi-grams or n-grams... those wouldn't be lexical models, but instead a separate 'layer' of language modeling.

Comment (mcdurdin) [reply common/predictive-text/package.json:2]: I don't think I was as clear on the distinction in my mind as you seem to be!

If that's the case, the lexical model compiler needs to be renamed to language-model-compiler. Because the majority of what it does is not specific to lexical models. Conceptually, they are paired. And I don't actually think we'd be developing a whole new set of types for other types of language models: the type data in `@keymanapp/lexical-model-types` is generic to any language model, I think.

Comment (jahorton) [reply common/predictive-text/package.json:2]: I was kind of thinking the same thing after I posted.  The one issue there - we actually do [publish that module](https://www.npmjs.com/package/@keymanapp/lexical-model-types).

I know there are ways to rename stuff within npm, so it can be fixed longterm once we're ready to do so.  But that'd take a bit of time and investigation, as I don't know the actual specifics yet... and that may break things for some people.  So... hmm.

File: .gitignore [hunk 13]
```diff
@@ -258,3 +258,5 @@ resources/environment.sh
 
 # Copy of the master VERSION.md used for Linux packaging
 /common/core/desktop/VERSION.md
+
+**/node_modules/
```
Comment (eddieantonio) [inline .gitignore:262]: This will work too:

```suggestion
node_modules/
```

File: common/predictive-text/package.json (hunk 10 above)
Comment (eddieantonio) [reply common/predictive-text/package.json:2]: Haha, I think we've named ourselves into a corner. For internal consistency, I vote for `@keymanapp/lexical-model-layer`, though I agree with Joshua that it is the _language modelling_ layer; not really the "lexical model" layer ¯\\\_(ツ)\_/¯
